module/dkg/recovery_test.go

```diff
@@ -0,0 +1,369 @@
+}
+
+// TestNewBeaconKeyRecovery_EpochIsNotCommitted tests a scenario:
+// - node is not in epoch committed phase
+// In a case like this there is no need to proceed since we don't have the next epoch available.
+func (s *BeaconKeyRecoverySuite) TestNewBeaconKeyRecovery_EpochIsNotCommitted() {
+	for _, phase := range []flow.EpochPhase{
+		flow.EpochPhaseFallback,
+		flow.EpochPhaseStaking,
+		flow.EpochPhaseSetup,
+	} {
+		s.currentEpochPhase = phase
+		recovery, err := NewBeaconKeyRecovery(unittest.Logger(), s.local, s.state, s.dkgState)
+		require.NoError(s.T(), err)
+		require.NotNil(s.T(), recovery)
+	}
+	s.dkgState.AssertNumberOfCalls(s.T(), "UpsertMyBeaconPrivateKey", 0)
+}
+
+// TestNewBeaconKeyRecovery_HeadException tests a scenario:
+// - exception is thrown when trying to get the head of the final snapshot
+// This is an unexpected error and should be propagated to the caller.
+func (s *BeaconKeyRecoverySuite) TestNewBeaconKeyRecovery_HeadException() {
+	exception := errors.New("exception")
+	s.finalSnapshot.On("Head").Unset()
+	s.finalSnapshot.On("Head").Return(nil, exception).Once()
+
+	recovery, err := NewBeaconKeyRecovery(unittest.Logger(), s.local, s.state, s.dkgState)
+	require.ErrorIs(s.T(), err, exception)
+	require.Nil(s.T(), recovery)
+}
+
+// TestNewBeaconKeyRecovery_EpochProtocolStateException tests a scenario:
+// - exception is thrown when trying to get the epoch protocol state of the final snapshot
+// This is an unexpected error and should be propagated to the caller.
+func (s *BeaconKeyRecoverySuite) TestNewBeaconKeyRecovery_EpochProtocolStateException() {
+	exception := errors.New("exception")
+	s.finalSnapshot.On("EpochProtocolState").Unset()
+	s.finalSnapshot.On("EpochProtocolState").Return(nil, exception).Once()
+
+	recovery, err := NewBeaconKeyRecovery(unittest.Logger(), s.local, s.state, s.dkgState)
+	require.ErrorIs(s.T(), err, exception)
+	require.Nil(s.T(), recovery)
+}
+
+// TestNewBeaconKeyRecovery_NextEpochCounterException tests a scenario:
+// - node is in epoch committed phase
+// - exception is thrown when trying to get counter of the next epoch
+// This is an unexpected error and should be propagated to the caller.
+func (s *BeaconKeyRecoverySuite) TestNewBeaconKeyRecovery_NextEpochCounterException() {
+	exception := errors.New("exception")
+	s.nextEpoch.On("Counter").Unset()
+	s.nextEpoch.On("Counter").Return(uint64(0), exception).Once()
+
+	recovery, err := NewBeaconKeyRecovery(unittest.Logger(), s.local, s.state, s.dkgState)
+	require.ErrorIs(s.T(), err, exception)
+	require.Nil(s.T(), recovery)
+}
+
+// TestNewBeaconKeyRecovery_NextEpochRetrieveMyBeaconPrivateKeyException tests a scenario:
+// - node is in epoch committed phase
+// - exception is thrown when trying to check if there is a safe beacon key for the next epoch
+// This is an unexpected error and should be propagated to the caller.
+func (s *BeaconKeyRecoverySuite) TestNewBeaconKeyRecovery_NextEpochRetrieveMyBeaconPrivateKeyException() {
```

Comment:
I would recommend to include another test like this one, with the only difference that https://github.com/onflow/flow-go/blob/eb968f879fa9085c798360f237597db944757147/module/dkg/recovery.go#L112 returns an `storage.ErrNotFound`. In this case, we would expect the code to proceed, right?